src/DnsClient/DnsTcpMessageHandler.cs

```diff
@@ -281,6 +290,30 @@ private async Task<DnsResponseMessage> QueryAsyncInternal(TcpClient client, DnsR
             return DnsResponseMessage.Combine(responses);
         }
 
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    foreach (var entry in _pools)
+                    {
+                        entry.Value.Dispose();
+                    }
+
+                    _pools.Clear();

```

Comment:
Not sure how would `ConcurrentDictionary` enumerator behave if it ends up racing with this `Clear()`. I would just remove it.